ports/analog/common-hal/busio/UART.c

```diff
@@ -0,0 +1,478 @@
+        MXC_UART_AbortAsync(self->uart_regs);
+        NVIC_DisableIRQ(MXC_UART_GET_IRQ(self->uart_id));
+        mp_raise_RuntimeError_varg(MP_ERROR_TEXT("UART read error"));
+    }
+
+    // Wait for transaction completion or timeout
+    while ((uart_status[self->uart_id] != UART_FREE) &&
+           (supervisor_ticks_ms64() - start_time < (self->timeout * 1000))) {
+    }
+
+    // If the timeout gets hit, abort and error out
+    if (uart_status[self->uart_id] != UART_FREE) {
+        MXC_UART_AbortAsync(self->uart_regs);
+        NVIC_DisableIRQ(MXC_UART_GET_IRQ(self->uart_id));
+        mp_raise_RuntimeError(MP_ERROR_TEXT("UART transaction timeout"));
+    }
+    // Check for errors from the callback
+    else if (uart_err != E_NO_ERROR) {
+        // todo: indicate error?
+        MXC_UART_AbortAsync(self->uart_regs);
+    }
+
+    // Copy the data from the ringbuf (or return error)
+    MXC_SYS_Crit_Enter();
+    err = ringbuf_get_n(context->ringbuf, data, len);
+    MXC_SYS_Crit_Exit();
+
+    return err;
+}
+
+// Write characters. len is in characters NOT bytes!
+// This function blocks until the timeout finishes
+size_t common_hal_busio_uart_write(busio_uart_obj_t *self,
+    const uint8_t *data, size_t len, int *errcode) {
+    int err;
+    uint32_t start_time = 0;
+    static size_t bytes_remaining;
+
+    // Setup globals & status tracking
+    uart_err = E_NO_ERROR;
+    uarts_active |= (1 << self->uart_id);
+    uart_status[self->uart_id] = UART_BUSY;
+    bytes_remaining = len;
+
+    mxc_uart_req_t uart_wr_req = {};
+
+    // Setup transaction
+    uart_wr_req.rxCnt = 0;
+    uart_wr_req.txCnt = 0;
+    uart_wr_req.rxData = NULL;
+    uart_wr_req.txData = data;
+    uart_wr_req.txLen = bytes_remaining;
+    uart_wr_req.rxLen = 0;
+    uart_wr_req.uart = self->uart_regs;
+    uart_wr_req.callback = (void *)uartCallback;
+
+    // Start the transaction
+    start_time = supervisor_ticks_ms64();
+    err = MXC_UART_TransactionAsync(&uart_wr_req);
+    if (err != E_NO_ERROR) {
+        *errcode = err;
+        MXC_UART_AbortAsync(self->uart_regs);
+        NVIC_DisableIRQ(MXC_UART_GET_IRQ(self->uart_id));
+        mp_raise_ValueError(MP_ERROR_TEXT("All UART peripherals are in use"));
+    }
+
+    // Wait for transaction completion or timeout
+    while ((uart_status[self->uart_id] != UART_FREE) &&
+           (supervisor_ticks_ms64() - start_time < (self->timeout * 1000))) {
+
+        // Call the handler and abort if errors
+        uart_err = MXC_UART_AsyncHandler(self->uart_regs);
+        if (uart_err != E_NO_ERROR) {
+            // todo: indicate error?
+            MXC_UART_AbortAsync(self->uart_regs);
+        }
+    }
+
+    // If the timeout gets hit, abort and error out
+    if (uart_status[self->uart_id] != UART_FREE) {
+        MXC_UART_AbortAsync(self->uart_regs);
+        NVIC_DisableIRQ(MXC_UART_GET_IRQ(self->uart_id));
+        mp_raise_RuntimeError(MP_ERROR_TEXT("Uart transaction timed out."));
```

Comment:
slight copy-editing
```suggestion
        mp_raise_RuntimeError(MP_ERROR_TEXT("UART transaction timed out"));
```
Q: what does this timeout mean? Other UART impls don't have write timeouts; they just hang. The `timeout` param for `busio.UART()` is documented as a read timeout. We don't say anything about write timeouts. Maybe we should. But I could only see a timeout if flow control is on and the write is held up due to that. There might be separate read and write timeouts if that were possible.